src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.CodeAnalysis.CSharp.Lowering
+{
+    class NullCheckRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        readonly MethodSymbol _method;
+        SyntheticBoundNodeFactory _fact;
+        public NullCheckRewriter(
+            MethodSymbol method,
+            SyntaxNode syntax,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            _method = method;
+            _fact = new SyntheticBoundNodeFactory(method, syntax, compilationState, diagnostics);
+        }
+
+        internal static BoundNode Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            int methodOrdinal,
+            VariableSlotAllocator slotAllocatorOpt,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics) //,
+                                       //out IteratorStateMachine stateMachineType)
+        {
+            var rewriter = new NullCheckRewriter(method, body.Syntax, compilationState, diagnostics);
+            return rewriter.Visit(body);
+        }
+
+        public override BoundNode VisitBlock(BoundBlock node)
+        {
+            List<SourceParameterSymbolBase> toCheck = new List<SourceParameterSymbolBase>();
+            for (int i = _method.Parameters.Length - 1; i >= 0; i--)
+            {
+                var param = (SourceParameterSymbolBase)_method.Parameters[i];
+                if (param.IsNullChecked)
+                {
+                    toCheck.Add(param);
+                }
+            }
+            return AddNullChecksToBody(node, toCheck);
+        }
+
+        private BoundNode AddNullChecksToBody(BoundBlock body, List<SourceParameterSymbolBase> toCheck)
+        {
+            BoundBlock prependedBody = body;
+            foreach (SourceParameterSymbolBase param in toCheck)
+            {
+                prependedBody = PrependBodyWithNullCheck(prependedBody, param);
+            }
+            return prependedBody;
+        }
+
+        private BoundBlock PrependBodyWithNullCheck(BoundBlock body, SourceParameterSymbolBase parameter)
+        {
+            if (parameter is null)
+                return body;
+
+            // If condition
```

Comment:
Consider replacing these comments with C# pseudocode similar to https://github.com/dotnet/roslyn/blob/aea735f367a5c6416686a4add7a81fdbf73af8a6/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs#L203